File: sdk-typescript/test/integ/execute_only_flow.ts
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

import {
  StepList,
  doubleCodec,
  goTo,
  gracefulComplete,
  type Context,
  type Flow,
  type Step,
  type StepDecision,
} from "../../src/index.js";

class ExecuteOnlySecondStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public getStepType(): string {
    return "ExecuteOnlySecondStep";
  }

  public execute(_context: Context, input: number): StepDecision {
    return gracefulComplete(input + 1);
  }
}

class ExecuteOnlyFirstStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(private readonly second: ExecuteOnlySecondStep) {}

  public getStepType(): string {
    return "ExecuteOnlyFirstStep";
  }

  public execute(_context: Context, input: number): StepDecision {
    return goTo(this.second, input + 1);
  }
}

export class ExecuteOnlyFlow implements Flow<number> {
  private readonly second = new ExecuteOnlySecondStep();
  private readonly first = new ExecuteOnlyFirstStep(this.second);

  public getFlowType(): string {
    return "ExecuteOnlyFlow";
  }

  public getSteps() {
    return StepList.startStep(this.first).otherSteps(this.second);
  }
}

File: sdk-python/tests/integ/execute_only_flow.py
# Portions of this file are derived from indeedeng/iwf-java-sdk.
# Those portions are licensed under the Apache License, Version 2.0.
# See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
#
# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications are licensed under the Super Durable Source License 1.0.
# Third-Party Materials remain under the Apache License, Version 2.0.
# See LICENSE and LEGACY_NOTICES.md.

from dex import Context, Flow, Step, StepDecision, StepList, go_to, graceful_complete


class ExecuteOnlySecondStep(Step[int]):
    def execute(self, context: Context, input: int) -> StepDecision:
        del context
        return graceful_complete(input + 1)


class ExecuteOnlyFirstStep(Step[int]):
    def __init__(self, second: ExecuteOnlySecondStep) -> None:
        self.second = second

    def execute(self, context: Context, input: int) -> StepDecision:
        del context
        return go_to(ExecuteOnlySecondStep, input + 1)


class ExecuteOnlyFlow(Flow[int]):
    def __init__(self) -> None:
        self.second = ExecuteOnlySecondStep()
        self.first = ExecuteOnlyFirstStep(self.second)

    def get_steps(self) -> StepList[int]:
        return StepList.start_step(self.first).other_steps(self.second)
